### ISSUE TITLE
RV vs object-wide logical clock

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#112684]
(followup from SIG API Machinery meeting of 2022-09-21)

(note, I'm keeping this text up to date so you don't have to read all the discussion below if you don't want to)

We are trying to unblock a class of caching behaviors, example PR: https://github.com/kubernetes/kubernetes/pull/112202

The problem is that today there is no way for a client to know which of two object states is "more recent". This makes it very hard to e.g. put the result of a write into a cache ahead of seeing the corresponding watch event.

(Aside: doing this destroys the property that listing out of cache always gives you a group of objects that actually co-existed in the server at some point in time. Not everyone needs this property, but we will probably ask that an in-tree assumption / write-through cache gives the option to restore this property. I'll refer to this as the "cross-object ordering problem".)

## Option 1: Relax prohibition on clients parsing resourceVersion.

Today we tell clients that they may compare RV as a string, only for equality. In practice, some clients actually do parse it anyway, and we already suffer the downside of not being able to change what we store in the string.

So this option is to tell clients that they may parse RV as a positive integer of up to 63 bits. The resulting integers may be compared to determine recency, but *only* between objects within the same collection (resource). RV comparison between collections is still invalid and at least some clusters are actually configured in a way that would break clients attempting this comparison.

Pros:
* People already do this. We lose almost no flexibility to change this in the future (we already lost it)
* Easy to see how this could solve the cross-object ordering problem (you can directly compare RV from objects in the same collection)
* No control plane changes needed, client changes compatible with old clusters.
 
Cons:
* Ugly; we would never do it this way if we started fresh
* 63 bits is slightly constraining for alternative DB implementations (but it's already that way)

## Option 2:
Alternatively, we could add a field to metadata, `.objectVersion`; this field would be a 63 bit int, a per-object logical clock. The server would enforce that it functions like the current `.generation` field, but for changes in the whole object, not just spec.

Pros:
* This design is much more aesthetic
* 63 bits is enough for a single object (especially since no-op changes won't increment this)
* [EDIT: perhaps in a few years we could start being more stringent about RV handling?]

Cons:
* Requires server changes, clients have to wait for a new version. ~Code has to take care for no-op updates.~ [EDIT: storage-layer changes aren't actually needed, only an api change]
* ~slight storage size increase.~ [EDIT: no, actually etcd already stores this for us]
* More difficult to solve the cross-object ordering problem (you have to wait for object X to hit at least objectVersion Y, instead of waiting to see any object with RV > Z)
* RV situation is probably not improved [EDIT: not in the short term at least].
* [EDIT: Awkward / not clear how to make this useful / play nice with other preconditions (including RV)]

## Recommendation

I recommend we select option 1, update documentation, and call it a day. [EDIT: but option 2 is back in the running after these updates...]

/sig api-machinery

**Comment by justinsb** [comment]
> We are trying to unblock a class of caching behaviors, example PR: https://github.com/kubernetes/kubernetes/pull/112202

Did we decide this was the right thing to do?  The alternative pattern that I've seen us adopt (I believe) is not to write to the cache, and simply to operate on changes as the cache observes them, waiting for changes to come in over the watch even if we made those changes ourselves.

**Comment by justinsb** [comment]
Also: We should check with all the aggregated apiserver implementations, as I think they are free to use strings if it makes them happy.  (Disclaimer: I'm working on an aggregated apiserver implementation in kpt/porch, and we're not doing resource-version right today, though I think/hope 63 bit ints should be enough for us)

**Comment by lavalamp** [comment]
> The alternative pattern that I've seen us adopt (I believe) is not to write to the cache, and simply to operate on changes as the cache observes them, waiting for changes to come in over the watch even if we made those changes ourselves.

You need one of these changes to make that safe, since you're not guaranteed to see every watch event.

**Comment by enj** [comment]
If we implement option 2 today, I feel like in a couple of years we could drop the "RV must be an int64" requirement (I have no evidence for this belief but parsing RV seems like uncommon behavior to me).

However, it is unclear to me how deletes would work with option 2 (since resetting the object version would be confusing to clients).

**Comment by lavalamp** [comment]
> how deletes would work with option 2 (since resetting the object version would be confusing to clients)

Clients already have to watch for UID changes. Yes, it would reset to zero.

**Comment by lavalamp** [comment]
> Also: We should check with all the aggregated apiserver implementations, as I think they are free to use strings if it makes them happy.

I would not say they are free to do that. I would say it's inadvisable, and clients that parse are rare enough that they might not discover the problem for a long time.

> (Disclaimer: I'm working on an aggregated apiserver implementation in kpt/porch, and we're not doing resource-version right today, though I think/hope 63 bit ints should be enough for us)

🙉

**Comment by stevekuznetsov** [comment]
@lavalamp are we more focused on solving the **intra**object ordering problem (for two snapshots in time of an object, which of $$a_0$$ and $$a$$ occurred first) or the **inter**object ordering problem (for two snapshots in time of two objects [of the same resource], which of $$a$$ and $$b$$ occurred first)? Your post seems to mention both, unless I'm reading it wrong.

Option 1 invites users to write logic that attempts to handle **inter**object ordering, but I fear that is (almost) always going to be based on incorrect assumptions - e.g. "this ServiceAccount token was written after the ServiceAccount was updated, so the token must be present" where the correct understanding must be to look at the observed generation/logical clock from the acting controllers? (I know SA/Secret are cross-resource, too, but even *that* would be invited by Option 1).

**Comment by lavalamp** [comment]
Both, but not everyone has the interobject ("cross-object" in the OP) ordering problem.

> Option 1 invites users to write logic that attempts to handle interobject ordering, but I fear that is (almost) always going to be based on incorrect assumptions - e.g. "this ServiceAccount token was written after the ServiceAccount was updated, so the token contents must be up-to-date" 

Hopefully people will realize that just because SOMETHING wrote the object, does not mean the controller of interest has updated it yet. Also we still forbid comparing across resource boundaries (e.g. secrets and service accounts).

**Comment by justinsb** [comment]
> You need one of these changes to make that safe, since you're not guaranteed to see every watch event.

I think the assumption in all of our controllers is that a watch sees _a_ state of all changed objects eventually, and in practice within a few seconds as long as the watch channel isn't overwhelmed.  I think a controller that performed the fast-forward optimization would actually behave _worse_ if we don't assume that, as the world would be even less self-consistent.

But anyway, I think I agree with the "OK fine, rv is an int63" suggestion here, I'm just not sure I agree with the linked issue as motivation.

I think we should think of this suggestion more in the context that @stevekuznetsov brought up though, and in particular whether we want to think about relaxing watch assumptions generally (for example, today we don't send duplicate create or update events, but doing so has blocked some optimizations I've wanted to make in the past), and I know that the cross-object-rv on LIST also prevents some optimizations (e.g. sharding by namespace)

**Comment by jpbetz** [comment]
I feel like option 1 is reasonable. I think offering a guarantee serves the community better at this point than holding on to option to change the behavior.

**Comment by lavalamp** [comment]
@justinsb the problem is with controllers that list all objects of some type and then do something like that. If one object has gone forward in time, some other number of objects have effectively gone back in time in comparison.

Like, imagine computing the free space on a Node by adding up all the pods on it; if one of your pods is from time T+3 and you haven't yet observed a pod deletion from T+2, you will add up the free space wrong and possibly evict a pod you needn't. It is really hard to debug and we don't want to do this to people.

**Comment by stevekuznetsov** [comment]
@lavalamp unless I'm missing something, option 2 does not address **inter**object comparisons, does it?

**Comment by lavalamp** [comment]
@stevekuznetsov It does (ish), see this bullet point:

> More difficult to solve the cross-object ordering problem (**you have to wait for object X to hit at least objectVersion Y, instead of waiting to see any object with RV > Z**)

**Comment by stevekuznetsov** [comment]
@lavalamp I'm sorry, I think I'm missing something here - if you have an independent logical clock per object, what mechanism is there to sync the timelines in order to allow causality comparisons across them?

**Comment by lavalamp** [comment]
@stevekuznetsov You have to assume that the watch events haven't been reordered.

**Comment by stevekuznetsov** [comment]
Today, if I have two `Secret`s I have no mechanism to tell which of them is "newer" than the other. The proposal in Option 2 is:

> Alternatively, we could add a field to metadata, .objectVersion; this field would be a 63 bit int, a per-object logical clock. The server would enforce that it functions like the current .generation field, but for changes in the whole object, not just spec.

So, now I have two `Secret`s and I can tell *for each of them* how many times its' spec and status have been changed. I can tell for any two snapshots of any individual `Secret` which is newer. How do I tell - given the per-object logical clocks - which of the two `Secret`s is newer?

**Comment by lavalamp** [comment]
You cannot in the abstract. I'm only seeking to solve this problem for people running a watch & a cache. (In fact it's almost a pro for option 2!)

**Comment by stevekuznetsov** [comment]
Gotcha. In my understanding at least, since you could not use the new field from Option 2 to take a list of `Secret`s and order them by recency, the option solves only the **intra**object problem, which is why my original question was asking which of the two problems was more pressing. The linked issue also only attempts to answer the **intra**object problem.

**Comment by lavalamp** [comment]
> The linked issue also only attempts to answer the intraobject problem.

Right, and this is a defect that needs to be fixed before we could merge it, see why in the example I gave in [this comment](https://github.com/kubernetes/kubernetes/issues/112684#issuecomment-1255525855).

**Comment by stevekuznetsov** [comment]
I see. To my eyes, Option 2 gives a cache author a simple way to provide either behavior (consistent or inconsistent collections), generalizes the `generation` concept clearly, and does not invite users to make broader, more potentially foot-gun-ey assumptions like **inter**object recency comparison outside of the tight constraints of LIST+WATCH. Tighter control over the API surface might be a valuable property as well for SIG APIMachinery.

**Comment by mm4tt** [comment]
Just FYI, implementing option 2 might be easier than you think. Etcd's KV has a concept of "version", which is exactly what you want to have IIUC - https://github.com/etcd-io/etcd/blob/main/api/mvccpb/kv.proto#L22

**Comment by stevekuznetsov** [comment]
@mm4tt the object's `resourceVersion` already exposes that field, just the API for Kubernetes doesn't allow users to parse it :)

**Comment by lavalamp** [comment]
Thanks @mm4tt, yeah that would make it super easy, I think.

**Comment by logicalhan** [comment]
I vote for 1.

**Comment by lavalamp** [comment]
> Is there an option 3? Could we define resource version as an unsized unsigned integer?

Unfortunately this would probably also break most or all clients parsing it, should we actually use more than 63 bits.

**Comment by smarterclayton** [comment]
> Unfortunately this would probably also break most or all clients parsing it, should we actually use more than 63 bits.

Ideally, if we wanted to "reclaim resourceVersion" we'd probably need to be smarter, as Daniel notes.  Provide a way for clients to get pre-broken as a feature gate, or per client, or opt-in.  All of those are expensive... and perhaps better addressed after we provide people a correct, relatively unobjectionable near term fix for intraobject.

You could argue that option 2 is the correct way to incrementally add a guarantee we did not actually intend to provide with rv, and once that has propagated we'd be in a spot to to formally limit what RV can be used for in a controlled way (by giving authors a replacement for the guarantee they took before).

Interobject can often (not always) be handled by describing the relationship explicitly using things like generation today (or a comparable resourceVersion / objectVersion), as folks note.

One problem neither of these options solve is violation of the clock, i.e. etcd restoration.  A reason I slightly prefer option 2 is that in the future a safely opaque RV can be mixed with the cluster ID of the etcd state, allowing client operations (watch, list, precondition edit) to be invalidated during a restore (failing to watch, forcing a list, or rejecting an edit).  That could be a very desirable incentive that cluster administrators could become our advocates are enforcing:

> If you want to correctly handle restores from etcd, you should set this config flag, which ensures clients reconnect properly during a restore.  Clients that parse resourceVersion may be broken in scenarios X, Y, and Z, and the solution is to use objectVersion instead, client libraries already support this.

If I were a distro looking to avoid incorrect behavior during an etcd restore, I would very much now have an incentive to go and fix old clients, and a community incentive to change.

**Comment by lavalamp** [comment]
> in the future a safely opaque RV can be mixed with the cluster ID of the etcd state, allowing client operations (watch, list, precondition edit) to be invalidated during a restore

Today you can accomplish this by incrementing the etcd global index prior to reconnecting the cluster. But yes, that could be cleaner.

(shall I put you down as voting for option 2 then?)

**Comment by enj** [comment]
> > Is there an option 3? Could we define resource version as an unsized unsigned integer?
> 
> Unfortunately this would probably also break most or all clients parsing it, should we actually use more than 63 bits.

In an abstract sense I was thinking:

1. Expand the schema of resource version (i.e. it can the current format plus some new format)
1. Document the full schema
1. Expand our libraries to support the new schema
1. Continue using the existing schema for N releases (minimum for N is probably 3)
1. At N+1, or some later release where we need the functionality, use the new schema

If the new schema is unsized unsigned integers, we would update our code to use big ints and document RV as unsized.

**Comment by lavalamp** [comment]
The problem with a discontinuous switch over is that you don't really have a good way of knowing if clients have been updated or not. I prefer a 2nd field for any sort of change (RV can get set to eg the large number mod 2^63).

**Comment by smarterclayton** [comment]
> Today you can accomplish this by incrementing the etcd global index prior to reconnecting the cluster

Right, if you know the increment of your lost control plane, which in a few cases has safety implications (if you lose two, you might not actually know how far ahead they were).

I would say I prefer option 2 because it fixes a concrete problem (intraobject) cleanly, is tractable for almost every storage system, and having solved it gives us more flexibility to tackle resourceVersion for cases we care about in our own code base plus a reasonable review of the ecosystem.

**Comment by lavalamp** [comment]
> Right, if you know the increment of your lost control plane, which in a few cases has safety implications (if you lose two, you might not actually know how far ahead they were).

(Getting off topic, but you can just add like 10 billion, 63 bits means you can waste some and be fine. Or you can look at client logs etc)

**Comment by deads2k** [comment]
I don't think we can ever practically change resourceVersion away from an integer.  Option 1 acknowledges that and provides the information we need.  I don't object to option 2 in concept, but if we acknowledge that we won't ever change resourceVersion away from an integer, it's unnecessary.  I don't think any deprecation period would be enough to make us confident in a switch.

**Comment by lavalamp** [comment]
I thought of another difficulty with option 2 last night... Do we treat it as a precondition instead of RV? In addition to RV? Or not at all? It seems any answer to this is pretty awkward.

**Comment by logicalhan** [comment]
> I don't think we can ever practically change resourceVersion away from an integer. 

My sentiments exactly.

**Comment by lavalamp** [comment]
Does anyone strongly object to option 1 here? It seems to be winning.

**Comment by stevekuznetsov** [comment]
> I thought of another difficulty with option 2 last night... Do we treat it as a precondition instead of RV? In addition to RV? Or not at all? It seems any answer to this is pretty awkward.

There is no precondition on `.metadata.generation`, so why would we necessarily need to change the precondition logic to be aware of the object version?

> Does anyone strongly object to option 1 here? It seems to be winning.

Has the implication of explicitly allowing comparability of recency between objects been thought through? It's one thing to say that there are bad actors parsing RV today, and another entirely to say that the surface expands to expose inter-object comparison.

**Comment by logicalhan** [comment]
> Has the implication of explicitly allowing comparability of recency between objects been thought through? It's one thing to say that there are bad actors parsing RV today, and another entirely to say that the surface expands to expose inter-object comparison.

Doesn't your cockroach DB implementation kind of depend on this?

**Comment by logicalhan** [comment]
> As of today, it is invalid to parse resource version.

That's what we say [not what we do](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/client-go/tools/cache/mutation_cache.go#L240-L262). 

> I also don't really understand what that topic has to do with this here.

It has a lot to do with it. If we commit to option 1, we commit to upholding fundamental assumptions underlying a bunch of alternate storage implementations (k3s comes to mind).

> I'm asking if the implications there have been thought through, it seems like a large departure from the history.

It's only a large departure from what we've admitted was allowed, but not a large departure from what has been in practice.

**Comment by stevekuznetsov** [comment]
> That's what we say [not what we do](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/client-go/tools/cache/mutation_cache.go#L240-L262).

That library is supposed to only be used in tests, and it's a [surprise](https://github.com/kubernetes/kubernetes/pull/112202#issuecomment-1255546834) even to @lavalamp that it is used elsewhere.

> It has a lot to do with it. If we commit to option 1, we commit to upholding fundamental assumptions underlying a bunch of alternate storage implementations (k3s comes to mind).

Neither option changes any fundamental assumptions, unless I'm missing something? 

> It's only a large departure from what we've admitted was allowed, but not a large departure from what has been in practice.

I'm not sure this is a super convincing argument - are there other examples of folks parsing RV? The implication from what you write is that it's common - is it? A test library never meant to be used in prod and accidentally used once doesn't seem to be enough supporting evidence there.

**Comment by smarterclayton** [comment]
> That's what we say [not what we do](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/client-go/tools/cache/mutation_cache.go#L240-L262).

I want to be clear that if we had caught this at the time it was introduced, it would have been blocked.  I didn't catch it, but I would absolutely prevent new code like this from going in.  I view that code as broken and would prefer a fix that allows us to address that.  If someone added something new I would force them to do something like `objectVersion` over option 1, although I understand why we are debating the pragmatism of option 1.

We made a deliberate decision to make resource version comparability across resource types invalid (to allow sharding by resource).  We made a deliberate decision to make resource version opaque (but didn't enforce it) to preserve future use extension.   Both of those are only enforced by documentation, but in practice are mostly honored.

Like all problems, we can't just force RV to be opaque quickly.  However, making it parseable has the same problem that the documentation about comparability does today - we can't restrict what is comparable without clearly having opacity - but we can't ever have opacity without providing a way to solve intraobject comparison today (option 2).

**Comment by smarterclayton** [comment]
Heh, I *think* this is the first reference to opaque resource version https://github.com/kubernetes/kubernetes/issues/1184#issuecomment-54554401 - will go do more dumpster diving later.

**Comment by smarterclayton** [comment]
One thing Han and I discussed (right at end) was that if we allow parsing of string we open our clients up to non-obvious immediate failures.  The most immediately obvious one is allowing a Javascript client to parse the string - you will get at most 52 bits of precision.  However, that's still probably close to larger than the biggest possible value.  But we should be careful about assuming 63 as the real max - especially if we're throwing around high bit twiddling or adding very large numbers.

Other hypothetical limitations - we can't ever export a cluster's contents from one storage engine to another without preserving the resource version / incrementing it IF you are allowing interobject references encoded with a resourceVersion.  I.e. we never really clarified how durable a resource version is - right now, it's within the lifespan of the cluster.  Some of those are guarantees I'd like for us to clarify if we retract our resourceVersion opacity statement.

Han made a good point that future non-etcd backends have already demonstrably shown they can implement total ordering by emulating RV, even if inefficiently in anything other than cockroachdb (due to the HLC impl).  Part of the original goal of shielding clients from etcd was to preserve flexibility of implementation - so option 1 would require us to give up some flexibility in theory (although not necessarily in practice).

I very much like finding ways to disallow comparison of resource versions across types (since documentation doesn't stop people), but otoh there are very interesting future use cases that potentially would be desirable in alternate implementations of storage such as "watch everything in a namespace" or "receive atomic schema change notifications in watch (for CRDs).

One disadvantage of allowing parseable resourceVersions - there could very well be aggregated API implementations that do not return integral resource versions, and we would break them by dictating clients may parse those RVs.  We're a little far in our lifecycle to make those assumptions.

**Comment by logicalhan** [comment]
> We made a deliberate decision to make resource version comparability across resource types invalid (to allow sharding by resource, which I know GKE uses and aggregate APIs might). 

Interestingly, I would reason in the other direction. Since we allow sharding by resource-type, it is therefore illegitimate to compare RVs across resource-types. 

> We made a deliberate decision to make resource version opaque (but didn't enforce it) to preserve future use extension. 

Giving people explicit guarantees that we're not going break ordering/comparable semantics (which are pretty much intrinsic now to Kubernetes and its protocols) is valuable. Especially considering the current state of etcd and considering the rampant community interest in alternative storage (crdb, k3s, etc.), it seems not only pragmatic but also prudent to offer explicit guarantees about RV. 

> there could very well be aggregated API implementations that do not return integral resource versions

My counterargument is that aggregated API servers *in general* are prone to errors around implementing Kubernetes protocols, since they simply have the burden of actually having to implement them.

**Comment by smarterclayton** [comment]
Option 2 is https://github.com/kubernetes/kubernetes/issues/7328

**Comment by smarterclayton** [comment]
> Especially considering the current state of etcd and considering the rampant community interest in alternative storage (crdb, k3s, etc.), it seems not only pragmatic but also prudent to offer explicit guarantees about RV.

Right.  We are proposing the degree of expansion of guarantees about RV we offer - we documented (but did not otherwise limit) a very restrictive definition.  Option 1 is "maximal expansion with no restriction".  Option 2 is "no expansion, add another field".

> My counterargument is that aggregated API servers in general are prone to errors around implementing Kubernetes protocols, since they simply have the burden of actually having to implement them.

A properly written aggregated API server may return non integral resource versions and be fully supported by our client library today. Encouraging clients to interpret those values more tightly pushes potential errors to the hardest place to find.  We generally do not tighten validation, and when we do we strive to do it as close to the source as possible.  An option that breaks potentially valid aggregated APIs for minimal benefit to clients (having to parse themselves) is generally against our API policy, although Daniel, David, and I have been actively arguing about the exact rule.

An option `1+2a` could be that the net new field is an integer and a scope they can be compared.  And then on the API server we make those `Resource` scoped by default and if someone comes up with a use case we add it.

**Comment by smarterclayton** [comment]
Heh, v1 metrics server (the podmetrics/nodemetrics adapters) returns empty string for resourceVersion.  I feel like I had a watch discussion with them at some point - although maybe that was prometheus.

**Comment by logicalhan** [comment]
> Heh, v1 metrics server (the podmetrics/nodemetrics adapters) returns empty string for resourceVersion. I feel like I had a watch discussion with them at some point - although maybe that was prometheus.

😂

**Comment by lavalamp** [comment]
It still seems like option 1 is winning to me, option 2 would be a fair amount more work, we'd need a volunteer etc...

**Comment by stevekuznetsov** [comment]
I'm more than happy to do the work. Is the argument for the first option is that it's worth it to impose that cost on all existing extension servers and expand the API surface to save the effort of plumbing an etcd field out through the API?

**Comment by lavalamp** [comment]
@stevekuznetsov No, I don't think anyone is saying that? Such aggregated apiservers are already broken if the right (wrong) client tries to talk to them. Formalizing the current situation doesn't make anyone any worse off as far as I can see.

**Comment by lavalamp** [comment]
IMO: since we have no formalized spec for us to make promises about, aggregated apiserver authors are signing up for this risk any time they deviate from what in-tree apiserver does. It's part of the reason why I will recommend not doing that if at all possible.

**Comment by stevekuznetsov** [comment]
@lavalamp I think I've lost a bit of context over the time that this has been discussed. Could you please help me understand the pro/con framing that is continually pushing you to Option 1?

**Comment by lavalamp** [comment]
@stevekuznetsov the pros & cons are listed in the OP, aside from effort which is massively tilted in favor of option 1.

**Comment by lavalamp** [comment]
All, I just posted #113056 which is about half of the documentation change needed to do option 1. Totally open to any suggestions on that to make option 1 easier for anyone impacted to deal with.

**Comment by stevekuznetsov** [comment]
@lavalamp I guess the cons for 2 are really not convincing whatsoever - 


- Requires server changes, clients have to wait for a new version.
This is plumbing an etcd field, fairly trivial and I expect also trivial to backport if needed.

- slight storage size increase. [EDIT: no, actually etcd already stores this for us]
As noted, not a con.

- More difficult to solve the cross-object ordering problem
You just have to wait for one integer to be larger than another integer, but instead of comparing to global RV you compare to the object-local number. Is this really "harder"?

- RV situation is probably not improved [EDIT: not in the short term at least].
A PR with option 2 + stronger wording about RV expectations could solve this the same way that a docs PR from option 1 does ...

- EDIT: Awkward / not clear how to make this useful / play nice with other preconditions (including RV)]
You still have not explained what you mean here, really. The value of the field from option 2 should be treated 100% identically to object generation. There is no awkwardness or ambiguity in my eyes - what are you worried about?

**Comment by stevekuznetsov** [comment]
@divanodestiny Option 1 would only require this invariant for each kind - as comparison across resource boundaries would still not be possible, so sharding *should* be possible per resource type. How does this inform your sharding story?

**Comment by divanodestiny** [comment]
> 



> @divanodestiny Option 1 would only require this invariant for each kind - as comparison across resource boundaries would still not be possible, so sharding _should_ be possible per resource type. How does this inform your sharding story?

We have implement a new backend instead of etcd with better performance and larger capacity. Our K8s clusters on the new background work well with more than 10K nodes and more than 300K nodes for years.  It help up reduce management cost and resource fragmentation. 

If Option 1 is adopt,  the revision of changes within a resource type should be a total ordered set but not a partially ordered set.  It prevents the backend sharding the data and event pipeline within a resource type. 

In my option, K8s is an event-driven distributed system.  I think in most cases we just aware the changes by events and do something according to the current state. If two changes happened in the same time, do we really needs to recognize the order of them?

**Comment by stevekuznetsov** [comment]
If you have time to join the SIG meeting today, it would be great to hear your opinion. As we found out in #109730  it's actually today required that there is a total order for events within a kind, to ensure that you can resume from any point. I'm wondering - if your backend does not provide a total order within a resource type, how do clients handle this edge case?

**Comment by divanodestiny** [comment]
> If you have time to join the SIG meeting today, it would be great to hear your opinion. As we found out in #109730 it's actually today required that there is a total order for events within a kind, to ensure that a client can resume a watch from any point. I'm wondering - if your backend does not provide a total order within a resource type, how do clients handle this edge case?

Sorry for the late response. 

We keep events in total order now. The largest cluster in prod env has 20k nodes and 1M pods now. But when we further increase the cluster size，the latency of event increase rapidly. We find there is a performance bottleneck on one node sorting the events. So we want to shard the revision like kcp in backend，it could involve some changes in ApiServer to event  that client resume watch and break the total order of event. The design is not complete now.

**Comment by stevekuznetsov** [comment]
@divanodestiny I see - since you can't issue cross-resource WATCH calls, you are actually free to shard by resource without any implication to the k8s API semantics. If you have 1M Pods, though, not sure that would help.

**Comment by liggitt** [comment]
The inability for clients to introspect resourceVersion means we have latent bugs in things like controller expectations that we cannot currently resolve - see https://github.com/kubernetes/kubernetes/issues/129795#issuecomment-2657716713
